dom/events/scrolling/scrollend-event-for-scroll-combination.html

```diff
@@ -0,0 +1,97 @@
+    <style>
+        #targetDiv {
+            width: 200px;
+            height: 200px;
+            overflow: scroll;
+        }
+
+        #innerDiv {
+            width: 400px;
+            height: 2000px;
+        }
+    </style>
+</head>
+<body style="margin:0" onload=runTest()>
+    <div id="targetDiv">
+        <div id="innerDiv">
+        </div>
+    </div>
+</body>
+
+<script>
+var target_div = document.getElementById('targetDiv');
+var user_scrollend_arrived = false;
+var programmatic_scrollend_arrived = false;
+var programmatic_scroll_done = false;
+var programmatic_scroll_top = 500;
+
+function onScroll(event) {
+    if (programmatic_scroll_done)
+        return;
+
+    // Perform programmatic scroll during user scrolling.
+    if (target_div.scrollTop > 50) {
+        target_div.scrollTop = programmatic_scroll_top;
+        programmatic_scroll_done = true;
+    }
+}
+
+function onScrollEnd(event) {
+    assert_false(event.cancelable);
+    assert_false(event.bubbles);
+    if (!programmatic_scrollend_arrived) {
+        programmatic_scrollend_arrived = true;
+        assert_equals(target_div.scrollTop, programmatic_scroll_top, "The first scrollend event is generated by programmatic scroll");
+        return;
+    }
+    user_scrollend_arrived = true;
+}
+
+function resetTargetScrollState() {
+    target_div.scrollTop = 0;
+    target_div.scrollLeft = 0;
+}
+
+function checkFinalPosition(target, position, pause_time_in_ms) {
+    return new Promise((resolve, reject) => {
+        step_timeout(() => {
+            resolve();
+            assert_equals(position.x, target.scrollLeft);
+            assert_equals(position.y, target.scrollTop);
+        }, pause_time_in_ms);
+    });
+}
+
+function runTest() {
+    promise_test(async (t) => {
+        resetTargetScrollState();
+        // Make sure that no scrollend event is sent to window.
+        window.addEventListener("scrollend",
+            t.unreached_func("window got unexpected scrollend event."));
+        await waitForCompositorCommit();
+        target_div.addEventListener("scrollend", onScrollEnd);
+        target_div.addEventListener("scroll", onScroll);
+
+        // Hit the scrollbar of target div and wait for target_div to get scrollend event.
+        var scrollbar_width = target_div.offsetWidth - target_div.clientWidth;
+        assert_true(scrollbar_width > 0, "This test requires scrollbar.");
+        var origin = { x: target_div.offsetWidth - scrollbar_width / 2, y: target_div.offsetHeight - 50 };
+        var delta = { x: 0, y: 0 };
+        await mouseActionsInTarget(target_div, origin, delta, 1000);
+        await waitFor(() => { return user_scrollend_arrived; },
+            'target_div did not receive two scrollend events after pressing scrollbar and performing programmatic scroll.');
+        assert_true(target_div.scrollTop > 0);
+        await checkFinalPosition(target_div, { x: target_div.scrollLeft, y: target_div.scrollTop }, 300);
```

Comment:
My expectation here is that if scrollend event is received the offsets are updated as well on the main thread. Do we need to wait for 300ms before checking?

Reply:
Yes, that's true. The intent is to make sure there's no more scroll position change after receiving user scroll scrollend event.

Reply:
Right, presumably line 92 makes sure that we have scrolled and line 93 is asserting that we don't continue to scroll after the scrollend event.